homeassistant/components/sensor/filter.py

```diff
@@ -277,6 +288,47 @@ def _filter_state(self, new_state):
         return filtered
 
 
+@FILTERS.register(FILTER_NAME_TIME_SMA)
+class TimeSMAFilter(Filter):
+    """Simple Moving Average (SMA) Filter.
+
+    The window_size is determined by time, and SMA is time weighted.
+    """
+
+    def __init__(self, window_size, precision, entity):
+        """Initialize Filter."""
+        super().__init__(FILTER_NAME_TIME_SMA, 0, precision, entity)
+        self._time_window = int(window_size.total_seconds())
+        self.last_leak = None
+        self.queue = deque()
+
+    def _leak(self):
+        """Remove timeouted elements."""
+        now = int(datetime.now().timestamp())
+
+        while self.queue:
+            timestamp, _ = self.queue[0]
+            if timestamp + self._time_window <= now:
+                self.last_leak = self.queue.popleft()
+            else:
+                break
+        return now
+
+    def _filter_state(self, new_state):
+        now = self._leak()
+        self.queue.append((now, float(new_state)))
+        moving_sum = 0
+        start = now - self._time_window
+        _, prev_val = self.last_leak or (0, float(new_state))
+
+        for timestamp, val in self.queue:
+            moving_sum += (timestamp-start)*prev_val
```

Comment:
What you're doing here is a lower bound of the moving average:

<img width="626" alt="screen shot 2018-03-12 at 16 57 50" src="https://user-images.githubusercontent.com/6833237/37294832-ff1662fc-2616-11e8-9ea1-ac408a173bfc.png">


Ideally we would instead be using the mean of `prev` and `val`:

<img width="627" alt="screen shot 2018-03-12 at 16 58 41" src="https://user-images.githubusercontent.com/6833237/37294726-ca4c2a98-2616-11e8-9c27-c5b9619812d2.png">

Reply:
I based the implementation in this paper:

http://www.eckner.com/papers/Algorithms%20for%20Unevenly%20Spaced%20Time%20Series.pdf

My implementation is that of SMA_last, what you propose is SMA_linear. Both are acceptable filters, yet different ones.

Reply:
Ok that is true; I don't see much value in this SMA_last method over the trapezoids though. The latter just seems quite a bit more accurate in most cases.

Reply:
I will probably PR a second filter based on SMA_linear, but the use cases are different:

**SMA_last**: discrete observation values 
- sensors that are event based

**SMA_linear**: discretely-observed continuous-time stochastic processes 
- sensors which are polled